File: src/router/package/lookup.ts
import type { NextFunction, Request, Response } from '@tinyhttp/app'
import { prisma } from 'database.js'

type LookupResponse = Response & {
	locals: {
		query: string;
	};
}

export function middleware(request: Request, response: LookupResponse, next: NextFunction) {
	const query = request.params.package
	if (!query) {
		return response.status(400)
			.json({
				message: '400 Bad Request',
				error: 'Missing URL parameter: \':package\'',
				date: new Date()
			})
	}

	response.locals.query = query.toString()
	next()
}

export async function handler(_request: Request, response: LookupResponse) {
	const { query } = response.locals
	const pkgs = await prisma.package.findMany({
		where: {
			package: query,
			isPruned: false
		},

		orderBy: [
			{
				isCurrent: 'desc'
			},
			{
				repositoryTier: 'asc'
			}
		]
	})

	if (pkgs.length === 0) {
		return response.status(404)
			.json({
				message: '404 Not Found',
				error: 'Package not found',
				date: new Date()
			})
	}

	return response.status(200)
		.json({
			message: '200 Successful',
			date: new Date(),
			count: pkgs.length,
			data: pkgs.map(data => {
				const entries = Object.entries(data)
					.filter(([key]) => key !== 'uuid' && key !== 'isPruned')

				return {
					...Object.fromEntries(entries),
					refs: {
						repo: `${$product.api_endpoint}/jailbreak/repository/${data.repositorySlug}`
					}
				}
			})
		})
}

File: api/src/routes/package/lookup.rs
use crate::prisma::package;
use crate::utility::{merge_json, tokio_run};
use crate::{db::prisma, utility::json_respond};

use prisma_client_rust::Direction;
use serde_json::{json, Value};
use tide::{
	Request, Result,
	StatusCode::{BadRequest, NotFound, Ok as OK},
};

pub async fn package_lookup(req: Request<()>) -> Result {
	let query = match req.param("package") {
		Ok(query) => query.to_string(),
		Err(err) => {
			println!("Error: {}", err);
			return Ok(json_respond(
				BadRequest,
				json!({
					"message": "400 Bad Request",
					"error": "Missing URL parameter: \':package\'",
					"date": chrono::Utc::now().to_rfc3339(),
				}),
			));
		}
	};

	let packages = tokio_run(async move {
		return prisma()
			.await
			.package()
			.find_many(vec![
				package::package::equals(query.to_string()),
				package::is_pruned::equals(false),
			])
			.order_by(package::is_current::order(Direction::Desc))
			.order_by(package::repository_tier::order(Direction::Asc))
			.with(package::repository::fetch())
			.exec()
			.await
			.unwrap();
	});

	if packages.len() == 0 {
		return Ok(json_respond(
			NotFound,
			json!({
				"message": "404 Not Found",
				"error": "Package not found",
				"date": chrono::Utc::now().to_rfc3339(),
			}),
		));
	}

	return Ok(json_respond(
		OK,
		json!({
			"message": "200 Successful",
			"date": chrono::Utc::now().to_rfc3339(),
			"count": packages.len(),
			"data": packages.iter().map(|package| {
				let slug = package.repository_slug.clone();
				return merge_json(package, json!({
					"refs": {
						"repo": format!("{}/jailbreak/repository/{}", env!("CANISTER_API_ENDPOINT"), slug)
					}
				}))
			}).collect::<Vec<Value>>(),
		}),
	));
}
